tests/yarn/pass_key_evaluator.py

```diff
@@ -0,0 +1,110 @@
+import random
+from typing import Tuple, Iterable, Optional
+
+from base_model import BaseModel
+
+
+class PassKeyEvaluator:
+    system = (
+        "There is an important pass key hidden inside a lot of irrelevant text. Find this key and memorize it. "
+        "I will quiz you about the key.\n")
+
+    garbage = "The grass is green. The sky is blue. The Sun is yellow. Here we go. There and back again.\n"
+
+    information = (
+        "The pass key is {pass_key}. Remember it. {pass_key} is the pass key.\n"
+    )
+
+    quiz_question = "What is the pass key? The pass key is: "
+
+    def __init__(self, model: BaseModel, seed: int = 42):
+        super().__init__()
+        self.model = model
+        self.rng = random.Random(seed)
+
+    def format_prompt(self, garbage_count: int,
+                      key_position: int) -> Tuple[str, str, int]:
+        """Generates a text file and inserts an execute line at a random position."""
+        assert (
+            0 <= key_position <= garbage_count
+        ), f"key_position={key_position}, garbage_count={garbage_count}"
+
+        garbage_prefix = "".join(self.garbage for _ in range(key_position))
+        garbage_suffix = "".join(self.garbage
+                                 for _ in range(garbage_count - key_position))
+
+        pass_key = f"%06d" % random.randrange(1000000)
+        information = self.information.format(pass_key=pass_key)
+
+        fragments = [
+            self.system,
+            garbage_prefix,
+            information,
+            garbage_suffix,
+            self.quiz_question,
+        ]
+
+        return (
+            "".join(fragments),
+            pass_key,
+            self.model.count_tokens(garbage_prefix),
+        )
+
+    def evaluate(self,
+                 max_tokens: int,
+                 resolution: int = 100,
+                 n: int = 10) -> Iterable[Tuple[int, int, int]]:
+        assert max_tokens > 0
+        assert resolution > 1
+
+        garbage_count = max_tokens // self.model.count_tokens(self.garbage)
```

Comment:
Reserve headroom, so the generated key can fit into the model's context window in all cases:
```py
garbage_count = (max_tokens - headroom) // self.model.count_tokens(self.garbage)
```